tracer/src/Datadog.Trace/RemoteConfigurationManagement/RemoteConfigurationManager.cs

```diff
@@ -0,0 +1,295 @@
+        }
+
+        public void UnregisterProduct(string productName)
+        {
+            _products.TryRemove(productName, out _);
+        }
+
+        private async Task Poll()
+        {
+            await SemaphoreSlim.WaitAsync().ConfigureAwait(false);
+
+            try
+            {
+                var request = BuildRequest();
+                var response = _remoteConfigurationApi.GetConfigs(request).Result;
+
+                if (response?.Targets?.Signed != null)
+                {
+                    ProcessResponse(response);
+                    _targetsVersion = response.Targets.Signed.Version;
+                }
+            }
+            catch (Exception e)
+            {
+                _lastPollError = e.Message;
+            }
+            finally
+            {
+                SemaphoreSlim.Release();
+            }
+        }
+
+        private GetRcmRequest BuildRequest()
+        {
+            var appliedConfigurations = _products.Values.SelectMany(pair => pair.AppliedConfigurations.Values).ToList();
+            var cachedTargetFiles = appliedConfigurations.Select(cache => new RcmCachedTargetFile(cache.Path.Path, cache.Length, cache.Hashes.Select(kp => new RcmCachedTargetFileHash(kp.Key, kp.Value)).ToList())).ToList();
+
+            return new GetRcmRequest(BuildRequestClient(), cachedTargetFiles);
+
+            RcmClient BuildRequestClient()
+            {
+                return new RcmClient(_id, _products.Keys.ToList(), BuildRcmClientTracer(), BuildRequestClientState());
+            }
+
+            RcmClientTracer BuildRcmClientTracer()
+            {
+                return new RcmClientTracer(_runtimeId, _tracerVersion, _service, _env, _appVersion);
+            }
+
+            RcmClientState BuildRequestClientState()
+            {
+                var configStates = appliedConfigurations.Select(cache => new RcmConfigState(cache.Path.Id, cache.Version, cache.Path.Product)).ToList();
+                return new RcmClientState(_rootVersion, _targetsVersion, configStates, _lastPollError != null, _lastPollError);
+            }
+        }
+
+        private void ProcessResponse(GetRcmResponse response)
+        {
+            var actualConfigPath =
+                response.ClientConfigs
+                        .Select(RemoteConfigurationPath.FromPath)
+                        .ToDictionary(path => path.Path);
+
+            var changedConfigurationsByProduct = GetChangedConfigurations()
+               .GroupBy(config => config.Path.Product);
+
+            foreach (var productGroup in changedConfigurationsByProduct)
+            {
+                var product = _products[productGroup.Key];
+
+                try
+                {
+                    var uuid = _service.ToUUID();
+                    var configurations =
+                        productGroup
+                           .Where(rcmConfig => rcmConfig.Path.Id == _service || rcmConfig.Path.Id == uuid)
```

Comment:
I thought that having the `Id` segment in the `path` be `service.ToUUID` is an implementation detail that is specific to Live Debugger (that you've already encapsulated inside `LiveDebuggerProduct`), so should this code be here? 

RCM RFC is pretty explicit that the `Config Id` part of the path "is not necessarily an UUID" which makes this more confusing.

Reply:
Yes, you are right, some code leftovers.
Cleaned up at dc949fdee1082930c7d7b873326289fa22ad1a6f